src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+                {
+                    cumsum += data[i];
+                    _cumSumList[i] = cumsum;
+                    _cumSumShift[i] = cumsum;
+                }
+                for (int i = n; i < length; ++i)
+                {
+                    _cumSumList[i] = (_cumSumList[i] - _cumSumShift[i - n]) / n;
+                }
+                for (int i = 1; i < n; ++i)
+                {
+                    _cumSumList[i] /= (i + 1);
+                }
+                return _cumSumList;
+            }
+
+            private double CalculateScore(double mag, double avgMag)
+            {
+                double safeDivisor = avgMag;
+                if (Math.Abs(safeDivisor) < _eps)
+                {
+                    safeDivisor = _eps;
+                }
+                return (Math.Abs(mag - avgMag) / safeDivisor);
+            }
+
+            private void GetExpectedValue(double[] values, double[][] results)
+            {
+                //Step 8: Calculate Expected Value
+                var exps = CalculateExpectedValueByFft(GetDeanomalyData(values, GetAnomalyIndex(results.Select(x => x[1]).ToArray())));
+
+                for (int i = 0; i < results.Length; ++i)
+                {
+                    results[i][3] = exps[i];
+                }
+            }
+
+            private void GetMargin(double[] values, double[][] results, double sensitivity)
+            {
+                //Step 8: Calculate Expected Value
+                var exps = CalculateExpectedValueByFft(GetDeanomalyData(values, GetAnomalyIndex(results.Select(x => x[1]).ToArray())));
+
+                //Step 9: Calculate Boundary Unit
+                var units = CalculateBoundaryUnit(values, results.Select(x => x[0] > 0).ToArray());
+
+                //Step 10: Calculate UpperBound and LowerBound
+                var margins = units.Select(x => CalculateMargin(x, sensitivity)).ToList();
+
+                for (int i = 0; i < results.Length; ++i)
+                {
+                    results[i][3] = exps[i];
+                    results[i][4] = units[i];
+                    results[i][5] = exps[i] + margins[i];
+                    results[i][6] = exps[i] - margins[i];
+                    //Step 11: Update Anomaly Score
+                    results[i][1] = CalculateAnomalyScore(values[i], exps[i], units[i], results[i][0] > 0);
+                }
+            }
+
+            private int[] GetAnomalyIndex(double[] scores)
+            {
+                List<int> anomalyIdxList = new List<int>();
+                for (int i = 0; i < scores.Length; ++i)
+                    if (scores[i] > _deanomalyThreshold)
+                    {
+                        anomalyIdxList.Add(i);
+                    }
+
+                return anomalyIdxList.ToArray();
+            }
+
+            private double[] GetDeanomalyData(double[] data, int[] anomalyIdxList)
+            {
+                double[] deAnomalyData = (double[])data.Clone();
```

Comment:
>double[] deAnomalyData = (double[])data.Clone(); [](start = 16, length = 48)

Can this be converted to a field similarly to the rest of the arrays this class uses?

Reply:
Fixed.

---
In reply to: [427869960](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427869960) [](ancestors = 427869960)